### PR TITLE
applet.interface.spi_flashrom: New applet that implements the Flashrom Serprog protocol

### PR DESCRIPTION
**Pull request by neuschaefer**
So far this is a draft, and will likely require improvements in a few ways; I'm thankful for any pointers in that regard.

I have however tested this applet with two different flash chips and it seems to work fine.

## TODO

- [x] ~~Upload serprog.h and serprog-protocol.txt to the archive repo~~
- [x] update pdm lock
- [x] remove usage of `match`
- [x] remove boilerplate code
- [x] flake8
- [x] enum style
- [x] byte encoding/decoding style
- [x] verify that `--freq NNNN` actually has an effect (I received a report to the contrary)

### CODE REVIEW

**Comment by whitequark** [comment]
This PR fails CI.

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    @classmethod
+    def add_build_arguments(cls, parser, access):
+        super().add_build_arguments(parser, access)
+
+        for pin in cls.__pins:
+            access.add_pin_argument(parser, pin, default=True)
+
+    @classmethod
+    def add_run_arguments(cls, parser, access):
+        super().add_run_arguments(parser, access)
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:172]: All of this is extraneous.

```suggestion
```

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    async def run(self, device, args):
+        spi_iface = await self.run_lower(SPISerprogApplet, device, args)
+        return SPISerprogInterface(spi_iface)
+
+    def build_subtarget(self, target, args):
+        return super().build_subtarget(target, args)
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:179]: Extraneous.
```suggestion
```

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                    pass
+        handle_fut = asyncio.ensure_future(handle())
+        await asyncio.wait([handle_fut], return_when=asyncio.FIRST_EXCEPTION)
+
+# -------------------------------------------------------------------------------------------------
+
+class SPISerprogAppletTestCase(GlasgowAppletTestCase, applet=SPISerprogApplet):
+    @synthesis_test
+    def test_build(self):
+        self.assertBuilds()
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:202]: I don't think you need an `assertBuilds` test if you add no gateware.

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        ServerEndpoint.add_argument(parser, "endpoint")
+
+    async def interact(self, device, args, iface):
+        endpoint = await ServerEndpoint("socket", self.logger, args.endpoint)
+        async def handle():
+            while True:
+                try:
+                    handler = SerprogCommandHandler(iface, endpoint, self.logger)
+                    await handler.handle_cmd()
+                except asyncio.CancelledError:
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:192]: I *think* this is the default behavior? What happens if you remove the `try/except`?

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                    pass
+        handle_fut = asyncio.ensure_future(handle())
+        await asyncio.wait([handle_fut], return_when=asyncio.FIRST_EXCEPTION)
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:195]: I think this is exactly the same as `await handle()`? So you can inline the entire loop into `interact()`.

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+
+class SPISerprogInterface:
+    def __init__(self, interface):
+        self.lower       = interface
+
+    async def write_read(self, sdata, rlen):
+        assert len(sdata) > 0
+        await self.lower.write(sdata, hold_ss=rlen > 0)
+        return await self.lower.read(rlen)
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:17]: Two lines between every top-level entity. (This and a few other extremely basic stylistic requirements is something I might add a flake8 job for...)

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+class SerprogCommandHandler:
+    # Serprog protocol constants, based on:
+    #
+    # - https://review.coreboot.org/plugins/gitiles/flashrom/+/refs/tags/v1.0.2/serprog.h
+    # - https://review.coreboot.org/plugins/gitiles/flashrom/+/refs/tags/v1.0.2/Documentation/serprog-protocol.txt
+    S_ACK               = 0x06
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:24]: I think this entire command list should be an enum, with the docstring pointing to these files.

File: software/glasgow/applet/interface/spi_serprog/__init__.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    assert len(PROGNAME) == 16
+
+    def __init__(self, interface, endpoint, logger):
+        self.interface = interface
+        self.endpoint  = endpoint
+        self.logger    = logger
+
+    async def get_u(self, size):
+        data = await self.endpoint.recv(size)
+        return sum([x << 8 * i for i, x in enumerate(data)])
```
Comment (whitequark) [inline software/glasgow/applet/interface/spi_serprog/__init__.py:72]: I would use `int.from_bytes`/`int.to_bytes`. Also, I would probably not add such a generic mechanism at all with a function per size when inlining it at every point of use would reduce the LOC count by like 20.

**Review by whitequark — CHANGES_REQUESTED** [review]

**Comment by whitequark** [comment]
Thanks for working on this! I think flashrom support could be quite useful, for the same reasons we have both `jtag-openocd` and other JTAG options.

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 3 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py:202]: good point

**Comment by neuschaefer** [comment]
@whitequark thanks for the review! I'll get back to it in detail later

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 4 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py]: It seems to fail in the same way (with or without try/except) when I inject a fault by unplugging the glasgow.

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 3 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py]: removed the whole testcase

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 2 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py]: done

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 1 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py]: done

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 4 above)
Comment (whitequark) [reply software/glasgow/applet/interface/spi_serprog/__init__.py]: I think that doesn't result in cancellation, but rather another exception being raised.

**Review by whitequark — COMMENTED** [review]
Code looks OK now, aside from style. (Please do fix that!)

I will merge this sometime soon when I find time to test it with a flash myself. I will probably reword the applet description a bit.

Would you like to become the code owner for the new applet?

**Comment by whitequark** [comment]
Also, just to be clear: I think flake8's default settings are a net negative, but there is a small subset that we do already require (but don't have written dwon anywhere) and so it's easier to automate that.

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 6 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py:18]: Double blank lines between toplevel items: done

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 5 above)
Comment (whitequark) [reply software/glasgow/applet/interface/spi_serprog/__init__.py:193]: Ah, right. If I remember correctly, just `await asyncio.ensure_future(handle())` will behave the way you want wrt cancellation.

**Comment by neuschaefer** [comment]
I don't fully understand the coroutine and exception stuff here, but upon further testing I discovered the following:

- Without the try/except block, the applet stops after the first client disconnects, which I found somewhat undesirable. With it, multiple consecutive connections are accepted, until the applet is terminated explicitly.
- Without spawning a coroutine for the handler, I have to press ^C twice to terminate the applet, and it prints a python backtrace. With the coroutine, one ^C terminates the applet cleanly.

So I reintroduced these two things in the last two commits

**Comment by neuschaefer** [comment]
> Would you like to become the code owner for the new applet?

If that simply entails that I'll be pinged on pull requests changing this applet. yes please. I can at least review the serprog-specific details.

**Comment by whitequark** [comment]
> If that simply entails that I'll be pinged on pull requests changing this applet.

Pretty much. You can add yourself to the CODEOWNERS file then. I think I'll also need to add you to the org.

**Comment by neuschaefer** [comment]
I thought about adding serprog.h and serprog-protocol.txt to the archive repo, but since the "vendor" in this case is another open source project, it doesn't seem so necessary

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 5 above)
Comment (neuschaefer) [reply software/glasgow/applet/interface/spi_serprog/__init__.py:193]: Unfortunately not

**Comment by neuschaefer** [comment]
> I will probably reword the applet description a bit.

The formatting is a bit broken too (in --help output), but I'll leave it as is for now, if you're going to reword the text anyway

**Comment by neuschaefer** [comment]
I can squash the commits that need squashing, when it's convenient for you

**Comment by whitequark** [comment]
I'm currently reviewing this applet locally. I'm making a number of changes, which I'll squash into the commit introducing the applet but describe here:
- Rename from `spi-serprog` to `spi-flashrom` (like `jtag-openocd`); it appears that "serprog" is just "flashrom's private protocol" rather than anything established.
- Rewrite the documentation to be more concise and closer to that of `jtag-openocd`.
- Add cross-links between `spi-flashrom` and `memory-25x` help text, to make the operator aware of the alternative in both cases.
- Remove "preview" flag from the applet since it seems reasonably mature to me.
- Add "hold" and "wp" pin assignments, which are necessary to read many flashes when they're in a clip because by default or chance the flash would end up either write-protected (if WP is active) or just non-communicative at all (if HOLD is active); I encountered this issue during testing. The `memory-25x` applet handles this.

**Review by whitequark — APPROVED** [review]
Thank you for your excellent work on this applet!

File: software/glasgow/applet/interface/spi_serprog/__init__.py (hunk 5 above)
Comment (whitequark) [reply software/glasgow/applet/interface/spi_serprog/__init__.py]: Yep you're right, I had to add this back. What an unintuitive library...

**Pull request merged**